### PR TITLE
Disable HTTP2 while proxying a "Connection: upgrade" request

### LINKED ISSUE #88782: Proxying websocket to http2 capable server failed with malformed response
**What happened**:

We are using the UpgradeAwareHandler from apimachery to proxy HTTP requests. If our target server is HTTP2 capable websockets will fail with a message like "malformed HTTP response "\x00\x00\x18\x04\x00\x00\x00\x00\x00\x00\x05\x00\x04\x00\x00\x00\x03\x00\x00\x00\xfa\x00\x06\x00\x10\x01@\x00\x04\x00\x04\x00\x00"

The reason for this is the proxy does a raw dial and the TLS connection negoiates http2 as the next proto but then the UpgradeAwareHandler writes a http/1.1 request.  A fix is in https://github.com/kubernetes/kubernetes/pull/88781

**Environment**:
apimachinery v0.17.2

### PR DESCRIPTION
**Pull request by ibuildthecloud**
When proxying connection upgrade requests, like websockets, we dial
the target and then manually write the http.Request to the wire,
bypassing the http.Client.  In this scenario we are by default using
HTTP/1.1 from both the client and to the target server we are proxying.
Because of this we must disable HTTP2 in the TLS handshake so that the
server does not think we are writing a HTTP2 request. We do this by
setting the TLSConfig.NextProtos field to "http/1.1".

fix #88782

```release-note
NONE
```

### CODE REVIEW

**Comment by ibuildthecloud** [comment]
Fixes https://github.com/kubernetes/kubernetes/issues/88782

**Comment by liggitt** [comment]
Thanks, can we exercise the bug and fix with a test?

**Comment by mikedanese** [comment]
Doesn't this diable http/2 for all proxied requests? Why not just use:

https://github.com/kubernetes/kubernetes/blob/aeb88b6ecd898bbb788a4629d5e0163172c6c85f/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L65-L67

And pass in a transport with http/2 disabled?

**Comment by ibuildthecloud** [comment]
@mikedanese This disables http2 for only times that we use GetDialer which is only when we are handling a connection upgrade.  Correct me if I'm wrong, but connection upgrade is a http/1.1 concept so it is required to disable http2.  If the user passes in a transport to disable http2 that will disable http2 for all proxied request.  We'd prefer to use http2 on http2 compatible requests. This approach disables http2 only when we know it is explicitly disallowed.

**Comment by ibuildthecloud** [comment]
@liggitt will add test.

**Comment by ibuildthecloud** [comment]
@liggitt Updated with tests.

**Comment by mikedanese** [comment]
Yes, we shouldn't be using http/2 for upgrade requests. I thought we hit it here:

https://github.com/kubernetes/kubernetes/blob/e865c0b21945ca1ad0792e89c13bf9ed074a3806/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L388

But that function is unused now. This is all pretty messy at this point and is due for some cleanup.

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go
Comment (mikedanese) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go]: The above block tries hard not to modify the tlsConfig that is passed in. Do we need to make a copy here?

**Comment by mikedanese** [comment]
Can we either:

* Document that DialURL won't negotiate http/2 with ALPN.
* Make DialURL private since we don't use it anywhere else.

Alternatively, we could probably fix this outside of dial url by making sure that DialForUpgrade only uses a transport that has http/2 disabled.

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go
Comment (ibuildthecloud) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go]: You're correct, I thought it was already copied or new at this point but I see I'm wrong.

**Comment by ibuildthecloud** [comment]
@mikedanese I added documentation to the DialURL method to describe it's behavior regarding http2 and http/1.1.

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go
Comment (mikedanese) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go]:
```suggestion
		"both client and server support http2, but force to http/1.1 for upgrade": {
```

Comment (mikedanese) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go]:
```suggestion
		"both client and server support http2, but force to http/1.1 for upgrade": {
```

Comment (ibuildthecloud) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go]: updated

Comment (ibuildthecloud) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go]: updated

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go
Comment (enj) [inline staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go]: Empty slice should return `true` as well right?

```go
	if len(nextProtos) == 0 {
		return true
	}
```

**Review by enj — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go
Comment (ibuildthecloud) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go]: I thought nextproto nil and empty slice were treated differently, but I can't find reference to that so I'm most likely wrong.

Comment (ibuildthecloud) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go]: updated

**Comment by caesarxuchao** [comment]
@mikedanese RE. https://github.com/kubernetes/kubernetes/pull/88781#issuecomment-594978958, are you saying the UpgradeAwareHandler.Dial is not being used anymore? If so, let's add a deprecation notice, and remove it in the next release. Otherwise it's not right to disable http/2 for UpgradeAwareHandler.Dial.

[update] in fact there is [no backward compatibility requirements](https://github.com/kubernetes/apimachinery#compatibility) in apimachinery, we can remove the method if it's not used.

**Comment by mikedanese** [comment]
> are you saying the UpgradeAwareHandler.Dial is not being used anymore?

Yup, I was kind of confused since it's the only thing that might dial for no upgrade requests so I deleted it and did a full compile. No usage apparently.

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go
Comment (mikedanese) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go]: https://tools.ietf.org/html/rfc7230#section-6.7

We upgrade http requests to websocket or spdy for streaming request like exec and logs. The headers communicates that you are switching protocols from http1/1 to X so when the upgrade is complete you can send whatever over it like it's just a tcp connection. Sending an upgrade request over http/2 is a connection error because underlying tcp connections are shared between requests. We can eventually migrate off of http1/1 altogether but we are waiting for go support to land. https://github.com/golang/go/issues/27244#issuecomment-489469332

**Comment by ibuildthecloud** [comment]
I removed UpgradeAwareHandler.Dial and renamed DialURL to be private.

**Comment by mikedanese** [comment]
Tentatively tagged for 1.18 since it fixes a bug and we haven't blown the final whistle yet.

File: staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go
Comment (caesarxuchao) [reply staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go]: Thanks. It's an http header. 
nit: should be "Connection: upgrade" according to the RFC.

**Comment by caesarxuchao** [comment]
This lgtm, too.

**Comment by caesarxuchao** [comment]
/test pull-kubernetes-e2e-gce-network-proxy

Run the tests since this is on the path of the apiserver network proxy. Though it won't pass until https://github.com/kubernetes/kubernetes/pull/88863 gets merged

**Comment by mikedanese** [comment]
@caesarxuchao still needs approval.

**Comment by caesarxuchao** [comment]
/approve
/hold

I want to make sure pull-kubernetes-e2e-gce-network-proxy passes before merging this.

This is a bug fix so I think it can merge after code freeze.

**Comment by caesarxuchao** [comment]
@mikedanese @ibuildthecloud feel free to remove the hold if pull-kubernetes-e2e-gce-network-proxy passes.

**Pull request merged**
